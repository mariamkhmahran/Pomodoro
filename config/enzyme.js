import { configure } from "enzyme";
import Adapter from "enzyme-react-adapter-future";

configure({ adapter: new Adapter() });
