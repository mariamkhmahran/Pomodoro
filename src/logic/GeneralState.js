import localforage from "localforage";

async function set(attr, value, callback) {
  localforage.setItem(attr, value, callback);
  return await get(attr);
}

async function get(attr, callback) {
  return await localforage.getItem(attr, callback);
}

async function getIndex(attr, item, func) {
  let array = await get(attr);
  let index = array.findIndex(func);
  return index;
}

async function getAll(callback) {
  let tasks = await get("tasks");
  let done = await get("done");
  let queue = await get("queue");
  let cycles = await get("cycles");

  if (callback) {
    Promise.all([tasks, done, queue, cycles]).then(
      callback({ tasks, queue, done, cycles })
    );
  }
  return { tasks, done, queue, cycles };
}

async function removeItem(attr, func) {
  let array = await get(attr);
  let index = array.findIndex(func);
  if (~index) {
    array.splice(index, 1);
    await set(attr, array);
    return array;
  }
}

async function addAtBeginning(attr, item) {
  let array = await get(attr);
  await localforage.setItem(attr, [item, ...array]);
}

async function addAtEnd(attr, item) {
  let array = await get(attr);
  await localforage.setItem(attr, [...array, item]);
}

async function updateItem(attr, item, func) {
  let array = await get(attr);
  let index = array.findIndex(func);
  array.splice(index, 1, item);
  set(attr, array);
}

export { set, get, getIndex, getAll, removeItem, addAtBeginning, addAtEnd, updateItem };
