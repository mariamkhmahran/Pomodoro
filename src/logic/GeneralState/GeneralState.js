import localforage from "localforage";

async function set(attr, value, callback) {
  await localforage.setItem(attr, value, callback);
  return value;
}

async function get(attr, callback) {
  return await localforage.getItem(attr, callback);
}

async function getIndex(attr, func) {
  let array = await get(attr);
  let index = array.findIndex(func);
  return index;
}

async function getAll(callback) {
  let tasks = await get("tasks");
  let done = await get("done");
  let queue = await get("queue");
  let cycles = await get("cycles");

  if (callback) {
    Promise.all([tasks, done, queue, cycles]).then(
      callback({ tasks, queue, done, cycles })
    );
  }
  return { tasks, done, queue, cycles };
}

async function removeItem(attr, func) {
  let array = await get(attr);
  let index = await getIndex(attr, func);
  if (~index) {
    array.splice(index, 1);
    await set(attr, array);
    return array;
  }
}

async function addAtBeginning(attr, item) {
  let array = await get(attr);
  await localforage.setItem(attr, [item, ...array]);
}

async function addAtEnd(attr, item) {
  let array = await get(attr);
  await localforage.setItem(attr, [...array, item]);
}

async function updateItem(attr, item, func) {
  let index = await getIndex(attr, func);
  let array;
  if (~index) {
    array = await get(attr);
    array.splice(index, 1, item);
    await set(attr, array);
  }
}

async function reOrder(queue, startIndex, endIndex) {
  if (startIndex > -1 && endIndex < queue.length) {
    let result = [...queue];
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);
    set("queue", result);
    return result;
  }
  return null;
}

export {
  set,
  get,
  getIndex,
  getAll,
  removeItem,
  addAtBeginning,
  addAtEnd,
  updateItem,
  reOrder
};
