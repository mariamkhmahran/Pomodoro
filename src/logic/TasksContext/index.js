export { default } from "./TasksContext";
export * from "./TasksContext";
