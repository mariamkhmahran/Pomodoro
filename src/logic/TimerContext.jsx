import React from "react";

const POMODORO_TIME = 0.025 * 60 * 1000;
export const Context = React.createContext();

class TimerContext extends React.Component {
  time = POMODORO_TIME;

  tick = () => {
    const { isTimerRunning, remainigTime } = this.state;
    if (isTimerRunning) {
      let currentTime = Date.now();
      if (remainigTime >= 0) {
        this.setState({
          remainigTime: this.time - (currentTime - this.startTime)
        });
        window.requestAnimationFrame(this.tick);
      } else {
        this.completeCycle();
      }
    }
  };

  completeCycle = () => {
    this.setState(
      () => ({
        remainigTime: POMODORO_TIME,
        isTimerRunning: false,
        endOfCycle: true
      }),
      () => {
        this.setState({ endOfCycle: false });
      }
    );
  };

  startTimer = () => {
    this.setState(
      () => ({
        isTimerRunning: true,
        endOfCycle: false
      }),
      () => {
        this.startTime = Date.now();
        window.requestAnimationFrame(this.tick);
      }
    );
  };

  stopTimer = () => {
    this.time = this.state.remainigTime;
    this.setState({
      isTimerRunning: false
    });
  };

  state = {
    remainigTime: POMODORO_TIME,
    startTimer: this.startTimer,
    stopTimer: this.stopTimer,
    isTimerRunning: false,
    endOfCycle: false
  };

  render() {
    return (
      <Context.Provider value={this.state}>
        {this.props.children}
      </Context.Provider>
    );
  }
}
export default TimerContext;
