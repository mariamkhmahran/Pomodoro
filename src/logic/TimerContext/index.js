export { default } from "./TimerContext";
export * from "./TimerContext";
