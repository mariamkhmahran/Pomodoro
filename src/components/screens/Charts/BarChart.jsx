import React from "react";
import { Bar } from "react-chartjs-2";
import * as GeneralState from "../../../logic/GeneralState";

var weekday = new Array(7);
weekday[0] = "Sunday";
weekday[1] = "Monday";
weekday[2] = "Tuesday";
weekday[3] = "Wednesday";
weekday[4] = "Thursday";
weekday[5] = "Friday";
weekday[6] = "Saturday";

var chartOptions = {
  maintainAspectRatio: false,
  scales: {
    xAxes: [
      {
        gridLines: {
          display: false,
          color: "white"
        },
        ticks: {
          beginAtZero: true,
          fontColor: "white"
        }
      }
    ],
    yAxes: [
      {
        gridLines: {
          display: false,
          color: "white"
        },
        ticks: {
          beginAtZero: true,
          display: false
        }
      }
    ]
  }
};

export default class BarChart extends React.Component {
  state = {
    data: {
      labels: [],
      datasets: [
        {
          label: "Cycles",
          backgroundColor: "#F36969",
          borderColor: "#F36969",
          data: []
        },
        {
          label: "Tasks",
          backgroundColor: "#8EB7EE",
          borderColor: "#8EB7EE",
          data: []
        }
      ]
    }
  };

  async componentDidMount() {
    let newData = await GeneralState.getChartData();
    let data = this.state.data;
    data.labels = newData.cycles.map(d => weekday[d.day]);
    data.datasets[0].data = newData.cycles.map(d => d.value);
    data.datasets[1].data = newData.tasks.map(d => d.value);
    this.setState({ data });

    this.props.onDidMount && this.props.onDidMount(this);
  }

  render() {
    return (
      <div
        style={{
          width: "90%",
          height: "100%"
        }}>
        <Bar
          id="chart"
          data={this.state.data}
          options={chartOptions}
          height={500}
          width={700}
          onChange={this.changeHandler}
        />
      </div>
    );
  }
}
