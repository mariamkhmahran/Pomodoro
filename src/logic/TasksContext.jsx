import React from "react";
import * as GeneralState from "src/logic/GeneralState";

export const ContextTasks = React.createContext();

class TimerContext extends React.Component {
  state = {
    tasks: [],
    queue: [],
    done: [],
    addTask: this.addTask,
    addCycle: this.addCycle,
    markAsDone: this.markAsDone
  };
  componentDidMount() {
    this.update();
  }
  update = async () => {
    let state = await GeneralState.getAll();
    this.setState(
      () => ({
        ...state,
        addTask: this.addTask,
        addCycle: this.addCycle,
        markAsDone: this.markAsDone
      }),
      () => {
        console.log(this.state);
      }
    );
  };

  addTask = async ({ name, cycles, deadline }) => {
    let id = "_" + Math.random().toString(36);
    let task = {
      id: id.substr(2, 9),
      name: name,
      cyclesDone: 0,
      estimatedTime: cycles,
      deadline: deadline,
      subTasks: [],
      done: false
    };
    let queue = GeneralState.get("queue");

    let promise = GeneralState.set("tasks", [task, ...this.state.tasks]);
    GeneralState.set("queue", [...queue, task]);

    Promise.all([promise]).then(() => {
      this.update();
      var modal = document.getElementById("modal");
      modal.style.display = "none";
    });
  };

  addCycle = async task => {
    task.cyclesDone += 1;

    let Promises = [
      GeneralState.updateItem("task", task, t => t.id === task.id),
      GeneralState.updateItem("queue", task, t => t.id === task.id)
    ];

    Promise.all(Promises).then(() => {
      this.update();
    });
  };

  markAsDone = async task => {
    task.done = true;
    task.subTasks.forEach(task => {
      task.done = true;
    });

    let Promises = [
      GeneralState.removeItem("tasks", t => t.id === task.id),
      GeneralState.removeItem("queue", t => t.id === task.id),
      GeneralState.addAtBeginning("done", task)
    ];

    Promise.all(Promises).then(() => {
      this.update();
    });
  };

  render() {
    return (
      <ContextTasks.Provider value={this.state}>
        {this.props.children}
      </ContextTasks.Provider>
    );
  }
}
export default TimerContext;
