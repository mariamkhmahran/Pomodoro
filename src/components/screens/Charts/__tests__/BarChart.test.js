import React from "react";
import { shallow } from "enzyme";
import BarChart from "../BarChart";
import * as GeneralState from "src/logic/GeneralState";

test("BarChart.render", () => {
  const wrapper = shallow(<BarChart />, { lifecycleExperimental: true });

  expect(wrapper).toMatchSnapshot();
});

test("BarChart.data", async () => {
  await GeneralState.log();
  const wrapper = shallow(<BarChart />, { lifecycleExperimental: true });
  let data = await GeneralState.get("chartData");
  await wait(100);

  expect(wrapper.instance().state.data.datasets[1].data).toMatchObject(
    data.tasks.map(t => t.value)
  );
});

function wait(ms) {
  return new Promise(res => setTimeout(res, ms));
}
