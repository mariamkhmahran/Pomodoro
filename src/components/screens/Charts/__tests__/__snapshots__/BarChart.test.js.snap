// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart.render 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "width": "90%",
    }
  }
>
  <Bar
    data={
      Object {
        "datasets": Array [
          Object {
            "backgroundColor": "#F36969",
            "borderColor": "#F36969",
            "data": Array [],
            "label": "Cycles",
          },
          Object {
            "backgroundColor": "#8EB7EE",
            "borderColor": "#8EB7EE",
            "data": Array [],
            "label": "Tasks",
          },
        ],
        "labels": Array [],
      }
    }
    height={500}
    id="chart"
    options={
      Object {
        "maintainAspectRatio": false,
        "scales": Object {
          "xAxes": Array [
            Object {
              "gridLines": Object {
                "color": "white",
                "display": false,
              },
              "ticks": Object {
                "beginAtZero": true,
                "fontColor": "white",
              },
            },
          ],
          "yAxes": Array [
            Object {
              "gridLines": Object {
                "color": "white",
                "display": false,
              },
              "ticks": Object {
                "beginAtZero": true,
                "display": false,
              },
            },
          ],
        },
      }
    }
    width={700}
  />
</div>
`;
